# <font color = 'orange'> Logistic Regression

---

### <font color = 'Blue'> Import library, load dataset and understand data 

In [1]:
from sklearn.datasets import load_iris

In [2]:
dataset = load_iris()

In [3]:
dataset

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

#### <font color = '#AA00FF'> Observation :
* We have a dictionary which contains data and information dataset. 

---

### <font color = 'Blue'> Explore dataset

In [4]:
dataset.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

#### <font color = '#AA00FF'> Observation :
* 'DESCR' contains the description of the dataset.
* 'target_names' contains dependent feature name and it's data is present inside 'target'.
* 'feature_names' contains independent feature name and it's data is present inside 'data'.

In [5]:
# We will check the description of dataset
print(dataset['DESCR'])

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

---

### <font color = 'Blue'> Create dataframe

In [6]:
import pandas as pd 
import numpy as np 

In [7]:
df = pd.DataFrame(dataset['data'],columns = dataset['feature_names'])
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [8]:
# target data is loaded into dataframe
df['target'] = dataset['target']

df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0

In [9]:
df.shape

(150, 5)

In [10]:
df['target'].unique()

array([0, 1, 2])

#### <font color = '#AA00FF'> Observation :
* As we are doing binary classification problem we need 2 target values but we have 3 target values.So, we take only 2 values and remove 1 value i.e here we will remove '2'.

In [11]:
df['target'] != 2

0       True
1       True
2       True
3       True
4       True
       ...  
145    False
146    False
147    False
148    False
149    False
Name: target, Length: 150, dtype: bool

In [12]:
# removing records whose target value is 2
df_copy = df[df['target'] != 2]

df_copy.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0

In [14]:
df_copy['target'].unique()

array([0, 1])

In [15]:
df_copy.shape

(100, 5)

---

### <font color = 'Blue'> 1. Separating independent and dependent feature

In [16]:
x = df_copy.iloc[:,:-1]

y = df_copy.iloc[:,-1]

In [17]:
print(x)
print()
print(y)

    sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                 5.1               3.5                1.4               0.2
1                 4.9               3.0                1.4               0.2
2                 4.7               3.2                1.3               0.2
3                 4.6               3.1                1.5               0.2
4                 5.0               3.6                1.4               0.2
..                ...               ...                ...               ...
95                5.7               3.0                4.2               1.2
96                5.7               2.9                4.2               1.3
97                6.2               2.9                4.3               1.3
98                5.1               2.5                3.0               1.1
99                5.7               2.8                4.1               1.3

[100 rows x 4 columns]

0     0
1     0
2     0
3     0
4     0
     ..
95 

---

### <font color = 'Blue'> 2. Train Test Split of data

In [18]:
from sklearn.model_selection import train_test_split

In [19]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size = 0.2,random_state = 42) 

In [20]:
print('Length of our dataset is',len(df_copy))

print('Length of our train data is',len(x_train))
# Test data is 20% of actual data
print('Length of our test datat is',len(x_test))

Length of our dataset is 100
Length of our train data is 80
Length of our test datat is 20


---

### <font color = 'Blue'> 3. Model Training

In [21]:
from sklearn.linear_model import LogisticRegression

In [22]:
# intially let's not consider any parameters
classification = LogisticRegression()

classification

LogisticRegression()

In [23]:
classification.fit(x_train,y_train)

LogisticRegression()

---

### <font color = 'Blue'> 4. Predication

In [24]:
y_pred = classification.predict(x_test)

y_pred

array([1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0])

---

### <font color = 'Blue'> 5. Confusion Metrics, Accuracy, Classification Report 

In [25]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

In [26]:
# y_test is actual data
# y_pred is predicated data

print('Confusion Metrics :')
print(confusion_matrix(y_test,y_pred))

Confusion Metrics :
[[12  0]
 [ 0  8]]


#### <font color = '#AA00FF'> Observation :
* As FP and FN is 0 it is a very good model.

In [27]:
accuracy = accuracy_score(y_test,y_pred) * 100
print(f'Accuracy : {accuracy} %')

Accuracy : 100.0 %


#### <font color = '#AA00FF'> Observation :
* As FP and FN is 0 Accuracy is 100%

In [28]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       1.00      1.00      1.00         8

    accuracy                           1.00        20
   macro avg       1.00      1.00      1.00        20
weighted avg       1.00      1.00      1.00        20



---

### <font color = 'Blue'> Cross Validation

#### CV is done by dividing training data internally into training and validation data and calculate accuracy

#### Now, We have to apply logistic regression with cross validation and obtain multiple accuracy within the triaining data by spliting training data into training data and validation data.

In [29]:
from sklearn.model_selection import KFold

In [30]:
# by default number of split is 5
cv = KFold()

cv

KFold(n_splits=5, random_state=None, shuffle=False)

#### <font color = '#AA00FF'> Observation :
* Cross Validation number of spilt will be 5.

In [31]:
from sklearn.model_selection import cross_val_score

In [32]:
# there are the scores all the cross validation 
cv_scores = cross_val_score(classification,x_train,y_train,scoring = 'accuracy',cv = cv)

cv_scores

array([1., 1., 1., 1., 1.])

#### <font color = '#AA00FF'> Observation :
* Here, for cros_value_score we give 'clasification' estimator then tell model to split data internally into train and validation using x_train , y_train then accuracy scoring is need then number of split is told by cv.

In [34]:
final_score = np.mean(cv_scores)

# Now, we got the final cross validation score by doing average of cross validatino score 
final_score

1.0

---

### <font color = 'Blue'> One more example on Cross Validation

In [35]:
## Lets see more complex data

# make a prediction with a multinomial logistic regression model
from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression

# define dataset
X, y = make_classification(n_samples=1000, n_features=10, n_informative=5, n_redundant=5, n_classes=2, random_state=1)

In [36]:
X,y

(array([[ 2.56999479, -0.13019997,  3.16075093, ..., -1.93094078,
          3.26130366,  2.05692145],
        [ 0.34129317,  2.51321418, -0.80416572, ...,  6.24734437,
         -1.92769365,  2.9503149 ],
        [ 2.27539972,  3.36561455,  0.17164362, ...,  2.74693781,
          0.13492444,  2.00339547],
        ...,
        [ 0.5234359 ,  1.90466429,  0.93243365, ...,  1.53945231,
          1.90646166,  1.99458587],
        [ 1.33747921,  3.25859684,  0.78792366, ...,  5.18788314,
         -0.82071083,  3.51411431],
        [-0.98534299,  0.83919047,  2.5820803 , ...,  3.04705685,
          0.66885641,  3.32838496]]),
 array([1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1,
        1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1,
        0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0,
        1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0,
        0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0,
   

In [37]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [38]:
complex_class_model = LogisticRegression(max_iter = 200)
# model training
complex_class_model.fit(X_train,y_train)

LogisticRegression(max_iter=200)

In [39]:
y_pred_complex = complex_class_model.predict(X_test)
# prediction
y_pred_complex

array([0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0,
       0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0,
       1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0,
       0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1,
       0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1,
       0, 1])

In [40]:
print(confusion_matrix(y_test,y_pred_complex))
print(accuracy_score(y_test,y_pred_complex))
print(classification_report(y_test,y_pred_complex))

[[78 13]
 [29 80]]
0.79
              precision    recall  f1-score   support

           0       0.73      0.86      0.79        91
           1       0.86      0.73      0.79       109

    accuracy                           0.79       200
   macro avg       0.79      0.80      0.79       200
weighted avg       0.80      0.79      0.79       200



#### <font color = '#AA00FF'> Observation :
* Here, we can see that there are more errors and accuracy is less i.e 79.

In [41]:
# Now, Can we improve accuracy using cross validation

cv = KFold(n_splits = 5)
cv_scores = cross_val_score(complex_class_model,X_train,y_train,scoring = 'accuracy',cv = cv)

cv_scores

array([0.80625, 0.78125, 0.79375, 0.8125 , 0.85625])

In [42]:
# mean of cross validation scores
final_scores = np.mean(cv_scores)

final_scores

0.8099999999999999

#### <font color = '#AA00FF'> Observation :
* For stack holder we have to tell we have got accuracy of maximum of 85% and minimum of 78%. We should avoid telling 80% is accuracy not a single value. 

#### Further this cross validation can be used to get best parameters

---